### PR TITLE
Reconcile existing resources in broker syncer

### LINKED ISSUE #85: Reconcile existing resources in the syncer
When syncing local -> remote, on startup, the syncer should delete remote resources that no longer exist locally.

### PR DESCRIPTION
**Pull request by tpantelis**
Added a `Reconcile` method to the resource syncer that takes a list of resources and, if a resource does not exist in the source Store cache, it is enqueued for deletion.

The broker syncer uses this method for reconciling in both sync directions. For `RemoteToLocal`, it lists the local resources obtained from the `LocalToRemote` syncer cache. It passes the list to the `Reconcile` method on the `RemoteToLocal` syncer which only checks for remote resources, ie those with a cluster ID label set. Similarly, for `LocalToRemote`, it lists the remote resources obtained from the `RemoteToLocal` syncer cache and checks for synced local resources.

There's special consideration if the source namespace is `NamespaceAll` for reconciling `LocalToRemote`. In this case, it needs to know the namespace of the originating local resource. To handle this, when syncing, it adds a label to store the originating namespace.

Fixes #85 
Related to submariner-io/submariner#1244

**Note**: this is initially a draft b/c I want to test it with _submariner_ and _lighthouse_  but I welcome reviews in the meantime.

### CODE REVIEW

File: pkg/syncer/broker/syncer.go [hunk 1]
```diff
@@ -293,6 +293,29 @@ func (s *Syncer) Start(stopCh <-chan struct{}) error {
 		}
 	}
 
+	lister := func(s syncer.Interface) []runtime.Object {
+		list, err := s.ListResources()
+		if err != nil {
```
Comment (mangelajo) [inline pkg/syncer/broker/syncer.go:298]: Should we skip the error if err is api.IsNotExist(err) ?

File: pkg/syncer/broker/syncer.go [hunk 2]
```diff
@@ -293,6 +293,29 @@ func (s *Syncer) Start(stopCh <-chan struct{}) error {
...
+			klog.Errorf("Unable to reconcile - error listing resources: %v", err)
+			return nil
+		}
+
+		return list
+	}
+
+	for i := 0; i < len(s.syncers); i += 2 {
```
Comment (mangelajo) [inline pkg/syncer/broker/syncer.go:306]: Ok, I checked that the resource syncers are ready at this point, because you wait for cache sync before exiting the syncer.Start(stopCh) function.

File: pkg/syncer/broker/syncer_test.go [hunk 3]
```diff
@@ -267,6 +322,89 @@ var _ = Describe("Broker Syncer", func() {
...
+			test.SetClusterIDLabel(resource, "remote")
+			initialLocalResources = append(initialLocalResources, resource)
+
+			copy := resource.DeepCopy()
+			copy.SetNamespace(config.BrokerNamespace)
+			initialBrokerResources = append(initialBrokerResources, copy)
+		})
+
+		It("should not delete it from the local datastore on reconciliation", func() {
+			time.Sleep(100 * time.Millisecond)
```
Comment (mangelajo) [inline pkg/syncer/broker/syncer_test.go:349]: I was thinking if there would be a better way for doing this, may be we can put all the sleep values under a constant at least, if we needed to bump them we could bump them all at once.

File: pkg/syncer/broker/syncer.go (hunk 1 above)
Comment (mangelajo) [reply pkg/syncer/broker/syncer.go:298]: Oh I see that's not possible, any error will come from WaitForCacheSync,.. ;)

**Review by mangelajo — APPROVED** [review]

**Comment by mangelajo** [comment]
Trying on submariner repo: https://github.com/submariner-io/submariner/pull/1270
Trying on the lighthouse repo: https://github.com/submariner-io/lighthouse/pull/507
Testing in operator (not really needed): submariner-io/submariner-operator#1258

**Comment by mangelajo** [comment]
Testing in lighthouse: https://github.com/submariner-io/submariner-operator/pull/1258

File: pkg/syncer/broker/syncer_test.go (hunk 3 above)
Comment (tpantelis) [reply pkg/syncer/broker/syncer_test.go:349]: This sleep isn't sensitive in that too small an interval might cause intermittent failures. I do this to give the code time to process it, such that if there's a bug and it does delete it, we can detect it. Otherwise, it may still exist b/c the code in question hasn't  executed yet.

**Comment by tpantelis** [comment]
@mangelajo Thanks for the test PRs. I also intend to add unit tests cases in _submariner_ and _lighthouse_.  The latter will also likely need additional reconciliation for the stand-alone resource syncers (eg `ServiceExport` -> `ServiceImport`).

**Comment by mangelajo** [comment]
Oh the lighthouse one is wrong, I threw it at the submariner-operator, I don't think we use it there. Let me try to throw one at lighthouse.

File: pkg/syncer/broker/syncer_test.go (hunk 3 above)
Comment (mangelajo) [reply pkg/syncer/broker/syncer_test.go:349]: True, I was wondering if we could wait on the reconcile process to have finished  (at that point we can check that nothing was deleted). But it's ok for now I think

**Review by mangelajo — APPROVED** [review]

File: pkg/syncer/broker/syncer_test.go [hunk 4]
```diff
@@ -89,19 +94,25 @@ var _ = Describe("Broker Syncer", func() {
...
+		})
+
+		It("should correctly sync to the broker datastore", func() {
 			test.VerifyResource(brokerClient, resource, config.BrokerNamespace, config.LocalClusterID)
+		})
 
-			Expect(localClient.ResourceInterface.Delete(resource.GetName(), nil)).To(Succeed())
-			test.AwaitNoResource(brokerClient, resource.GetName())
+		Context("and then deleted", func() {
+			It("should be deletes from the broker datastore", func() {
```
Comment (sridhargaddam) [inline pkg/syncer/broker/syncer_test.go:108]: nit: Can be updated in a different PR 
```suggestion
			It("should be deleted from the broker datastore", func() {
```

**Review by sridhargaddam — APPROVED** [review]
LGTM, thanks @tpantelis

File: pkg/syncer/resource_syncer.go [hunk 5]
```diff
@@ -40,11 +40,15 @@ import (
 	"k8s.io/klog"
 )
 
+const OrigNamespaceLabelKey = "submariner-io/originatingNamespace"
```
Comment (vthapar) [inline pkg/syncer/resource_syncer.go:43]: We have something similar that we use in lighthouse for namespace as well as cluster. Now that admiral tags, we should cleanup labels in lighthouse.

**Review by vthapar — APPROVED** [review]

File: pkg/syncer/broker/syncer_test.go (hunk 3 above)
Comment (tpantelis) [reply pkg/syncer/broker/syncer_test.go:349]: yeah we'd have to somehow provide a chan for the tests to wait on. But even that wouldn't be complete as the delete is enqueued and processed on the workqueue.

File: pkg/syncer/broker/syncer_test.go (hunk 4 above)
Comment (tpantelis) [reply pkg/syncer/broker/syncer_test.go:108]: arghhh - I missed that ☹️

File: pkg/syncer/resource_syncer.go (hunk 5 above)
Comment (tpantelis) [reply pkg/syncer/resource_syncer.go:43]: yes we could although admiral only tags it for a particular case but we could change that.
